text/3502-cargo-script.md

```diff
@@ -0,0 +1,1229 @@
+- Files are considered to have embedded manifest if they end with `.rs` or they lack and extension and are of type file
+- This allows running `cargo test --manifest-path single.rs`
+- `cargo add` and `cargo remove` may not support editing embedded manifests initially
+- Path-dependencies may not refer to single-file packages at this time (they don't have a `lib` target anyways)
+
+Single-file packages will not be accepted as `path` or `git` dependencies.
+
+The lockfile for single-file packages will be placed in `CARGO_TARGET_DIR`.  In
+the future, when workspaces are supported, that will allow a user to have a
+persistent lockfile.
+We may also allow customizing the non-workspace lockfile location in the [future](#future-possibilities).
+
+## `cargo <file>.rs`
+
+`cargo` is intended for putting in the `#!` for single-file packages:
+```rust
+#!/usr/bin/env cargo
+
+fn main() {
+    println!("Hello world");
+}
+```
+- Like with `cargo install`, `.cargo/config.toml` will be read based on the
+  scripts location rather than the current-dir.
+  - And like `cargo install`, the current-dir `rust-toolchain.toml` is respected
+- `--release` is not passed in because the primary use case is for exploratory
+  programming, so the emphasis will be on build-time performance and debugging,
+  rather than runtime performance
+
+Most other flags and behavior will be similar to `cargo run`.
+
+The precedence for `cargo foo` will change from:
+1. built-in commands
+2. user aliases
+3. third-party commands
+
+to:
+1. built-in command xor manifest
+2. user aliases
+3. third-party commands
+
+To allow the xor, we enforce that
+- manifests must be passed in as `Cargo.toml`, with a `.rs` extension, or have a `/` in them
+  - So `./build` can be used to run a script name `build` rather than the `cargo build` command
+- no built-in command may look like an accepted manifest
+
+When piping `cargo <file>.rs`, `--quiet` will be assumed.
+Further work may be done to refine the output in interactive mode.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+The implicit content of the manifest will be unclear for users.
+We can patch over this as best we can in documentation but the result won't be
+ideal.
+A user can workaround this with `cargo metadata --manifest-path <file>.rs`
+or `cargo read-manifest --manifest-path <file>.rs`
+
+Users may be using this feature for heavier duty scripting and want optimized builds.
+- This can be worked around by changing `profile.dev` in the embedded manifest
+
+Like with all cargo packages, the `target/` directory grows unbounded.
+This is made worse by them being out of the way and the scripts are likely to be short-lived.
+Some prior art include a cache GC but that is also to clean up the temp files
+stored in other locations
+(our temp files are inside the `target/` dir and should be rarer).
+A GC for cargo is being tracked in [rust-lang/cargo#12633](https://github.com/rust-lang/cargo/issues/12633)
```

Comment:
it looks like the "Dealing with leaked `target/`" section confirms that `cargo clean` will work, but it might be worth explicitly calling this out here :)

Reply:
Mentioned in 99d5ea32